packages/common/src/PredicateConnector.ts

```diff
@@ -157,12 +160,25 @@ export abstract class PredicateConnector extends FuelConnector {
 
     const transactionRequest = transactionRequestify(transaction);
     const newestPredicate = this.getNewestPredicate();
-    if (!!newestPredicate && this.predicateAddress !== newestPredicate.key) {
-      const coinInput = transactionRequest.getCoinInputs();
-      transactionRequest.addChangeOutput(
-        Address.fromAddressOrString(newestPredicate.key),
-        coinInput[0]?.assetId as BytesLike,
-      );
+    if (!!newestPredicate && !this.predicateAccount.equals(newestPredicate)) {
+      const coinOutput = transactionRequest.getCoinOutputs();
+      const predicateAddress =
+        newestPredicate.getPredicateAddress(walletAccount);
+
+      // If there are no outputs, add a change output to the newest predicate
+      if (transactionRequest.getChangeOutputs().length === 0) {
+        transactionRequest.addChangeOutput(
+          Address.fromString(predicateAddress),
+          coinOutput[0]?.assetId as BytesLike,
+        );
+      }
+      // If there are outputs, change output addresses to the newest predicate
+      else {
+        transactionRequest.outputs.forEach((output) => {
+          if (output.type !== OutputType.Change) return;
+          output.to = Address.fromAddressOrString(predicateAddress).toB256();
+        });
+      }
```

Comment:
we can just replace the current change outputs IF it's equal to one of the predicate versions

we don't wanna add outputs that were not in tx before and also dont wanna replace outputs from another addresses that are not from predicate versions we know

Reply:
if there's no change outputs, we should not add it

Reply:
got it